src/auxiliary/mpi_arrays.jl

```diff
@@ -0,0 +1,228 @@
+# invalidations. Alternatively, could implement our own norm and pass that as
+# `internalnorm`. We could also try to use the least intrusive approach for now
+# and specialize `length` to return a global length while making sure that all
+# local behavior is still working as expected (if using `eachindex` instead of
+# `1:length` etc.). This means that we would have
+# `eachindex(u) != Base.OneTo(length(u))` for `u::TrixiMPIArray` in general,
+# even if `u` and its underlying array use one-based indexing.
+# Some consequences are that we need to implement specializations of `show`,
+# since the default ones call `length`. However, this doesn't work if not all
+# ranks call the same method, e.g., when showing an array only on one rank.
+# Moreover, we would need to specialize `copyto!` and probably many other
+# functions. Since this can lead to hard-to-find bugs and problems in MPI code,
+# we use a more verbose approach. Thus, we let `length` be a local `length` and
+# provide a new function `Trixi.ode_norm` to be passed as `internalnorm` of
+# OrdinaryDiffEq's `solve` function.
+
+
+# Specialization of `view`. Without these, `view`s of arrays returned by
+# `wrap_array` with multiple conserved variables do not always work...
+# This may also be related to the use of a global `length`?
+Base.view(u::TrixiMPIArray, idx::Vararg{Any,N}) where {N} = view(parent(u), idx...)
+
+
+"""
+    ode_norm(u, t)
+
+Implementation of the weighted L2 norm of Hairer and Wanner used for error-based
+step size control in OrdinaryDiffEq.jl. This function is aware of
+[`TrixiMPIArray`](@ref)s, handling them appropriately with global MPI
+communication.
+
+You must pass this function as keyword argument
+`internalnorm=Trixi.ode_norm`
+of `solve` when using error-based step size control with MPI parallel execution
+of Trixi.jl.
+"""
+ode_norm(u, t) = @fastmath abs(u)
+ode_norm(u::AbstractArray, t) = sqrt(sum(abs2, u) / length(u))
+function ode_norm(u::TrixiMPIArray, t)
+  local_sumabs2 = sum(abs2, parent(u))
+  local_length  = length(parent(u))
+  # TODO: MPI. This could be fused into one call to improve parallel performance.
+  global_sumabs2 = MPI.Allreduce(local_sumabs2, +, mpi_comm(u))
+  global_length  = MPI.Allreduce(local_length,  +, mpi_comm(u))
+  return sqrt(global_sumabs2 / global_length)
+end
+
+
+"""
+    ode_unstable_check(dt, u, semi, t)
+
+Implementation of the basic check for instability used in OrdinaryDiffEq.jl.
+Instead of checking something like `any(isnan, u)`, this function just checks
+`isnan(dt)`. This helps when using [`TrixiMPIArray`](@ref)s, since no additional
+global communication is required and all ranks will return the same result.
+
+You should pass this function as keyword argument
+`unstable_check=Trixi.ode_unstable_check`
+of `solve` when using error-based step size control with MPI parallel execution
+of Trixi.jl.
+"""
+ode_unstable_check(dt, u, semi, t) = isnan(dt)
```

Comment:
This fundamentally changes the way the error-based step size control works in parallel vs. in serial, doesn't it? That is, it makes the MPI version less sensitive and delay reactions to NaN values in the solution than in the serial version. Does this only affect some post-hoc instability check (i.e., to abort if the solution has already gone bad) or is it used for active step size control? In the latter case, I assume we would have to implement this in parallel, wouldn't we?

Reply:
Thanks a lot for the clarification!